discovery/ionos/server.go

```diff
@@ -0,0 +1,170 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ionos
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	ionoscloud "github.com/ionos-cloud/sdk-go/v6"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	serverLabelPrefix = metaLabelPrefix + "server_"
+
+	serverAvailabilityZoneLabel = serverLabelPrefix + "availability_zone"
+	serverBootCDROMIDLabel      = serverLabelPrefix + "boot_cdrom_id"
+	serverBootImageIDLabel      = serverLabelPrefix + "boot_image_id"
+	serverBootVolumeIDLabel     = serverLabelPrefix + "boot_volume_id"
+	serverCPUFamilyLabel        = serverLabelPrefix + "cpu_family"
+	serverDatacenterIDLabel     = serverLabelPrefix + "datacenter_id"
+	serverIDLabel               = serverLabelPrefix + "id"
+	serverIPLabel               = serverLabelPrefix + "ip"
+	serverLANIDLabel            = serverLabelPrefix + "lan_id"
+	serverLifecycleLabel        = serverLabelPrefix + "lifecycle"
+	serverNameLabel             = serverLabelPrefix + "name"
+	serverNICIDLabel            = serverLabelPrefix + "nic_id"
+	serverPrimaryIPLabel        = serverLabelPrefix + "primary_ip"
+	serverPrimaryLANIDLabel     = serverLabelPrefix + "primary_lan_id"
+	serverPrimaryNICIDLabel     = serverLabelPrefix + "primary_nic_id"
+	serverStateLabel            = serverLabelPrefix + "state"
+	serverTypeLabel             = serverLabelPrefix + "type"
+)
+
+type serverDiscovery struct {
+	*refresh.Discovery
+	client       *ionoscloud.APIClient
+	port         int
+	datacenterID string
+}
+
+func newServerDiscovery(conf *SDConfig, logger log.Logger) (*serverDiscovery, error) {
+	d := &serverDiscovery{
+		port:         conf.Port,
+		datacenterID: conf.DatacenterID,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "ionos_sd")
+	if err != nil {
+		return nil, err
+	}
+
+	// Username, password and token are set via http client config.
+	cfg := ionoscloud.NewConfiguration("", "", "", conf.ionosEndpoint)
+	cfg.HTTPClient = &http.Client{
+		Transport: rt,
+		Timeout:   time.Duration(conf.RefreshInterval),
+	}
+	cfg.UserAgent = fmt.Sprintf("Prometheus/%s", version.Version)
+
+	d.client = ionoscloud.NewAPIClient(cfg)
+
+	return d, nil
+}
+
+func (d *serverDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	api := d.client.ServersApi
+
+	servers, _, err := api.DatacentersServersGet(ctx, d.datacenterID).
```

Comment:
Does this use pagination underneat? Can we ask directly for many servers?

Reply:
The [`Servers` struct](https://pkg.go.dev/github.com/ionos-cloud/sdk-go/v6#Servers) includes attribute for [`PaginationLinks`](https://pkg.go.dev/github.com/ionos-cloud/sdk-go/v6@v6.0.5851#PaginationLinks) which can be used to get the next and previous pagination link. 

But, I haven't seen a method provided by the SDK for using the next pagination link. The Servers struct contains an `Offset` and `Limit` attribute, but these only seem to be set when explicitly specified in the API request call. The documentation on the [`DatacentersServersGet` method](https://pkg.go.dev/github.com/ionos-cloud/sdk-go/v6@v6.0.5851#ServersApiService.DatacentersServersGet) says that it returns all servers.

Furthermore, I've taken a look into the official Terraform provider provided by IONOS and they don't seem to perform pagination themselves. https://github.com/ionos-cloud/terraform-provider-ionoscloud/blob/f98403465589e1c3daa7af065db65a4bcaf488e2/ionoscloud/data_source_server.go#L663

So, I'm not sure. My hunch is that it's handled by the SDK or API.

Reply:
It's not done by the SDK, but I am fine with doing this as a follow-up PR.